### ISSUE TITLE
Multisampling realisation

### ISSUE THREAD

**Issue by Ilyabasharov** [SuLvXiangXin/zipnerf-pytorch#21]
Hello, @SuLvXiangXin ! Thank you for your work. I see your multisampling [code ](https://github.com/SuLvXiangXin/zipnerf-pytorch/blob/b1cb42943d244301a013bd53f9cb964f576b0af4/internal/render.py#L92) and I have some questions.

1) What is `cam_dirs`? I see this defenition on [code](https://github.com/SuLvXiangXin/zipnerf-pytorch/blob/b1cb42943d244301a013bd53f9cb964f576b0af4/internal/datasets.py#L385) but it gives me zero ideas why did you use `cam_dirs` in multisampling.
2) If we want to constract two vectors perpendicular to the view direction, we can use direction vector instead of `cam_dirs`, e.g
```python3
# two basis in parallel to the image plane
rand_vec = torch.randn_like(directions)
ortho1 = F.normalize(torch.cross(directions, rand_vec, dim=-1), dim=-1)
ortho2 = F.normalize(torch.cross(directions, ortho1, dim=-1), dim=-1)
```
Please correct me if I'm not right. Thanks!

**Comment by SuLvXiangXin** [comment]
@Ilyabasharov You can have a look at the MipNeRF paper. They truncate the ray cone parallel to the image plane, rather than perpendicular to the view direction. I think ZipNeRF may adopt this configuration.

**Comment by Ilyabasharov** [comment]
@SuLvXiangXin do you mean this? 4 page of MipNeRF paper
![github](https://github.com/SuLvXiangXin/zipnerf-pytorch/assets/25929618/a8ed3c3b-8281-48ad-819d-2c63602db5a1)

**Comment by SuLvXiangXin** [comment]
<img width="525" alt="image" src="https://github.com/SuLvXiangXin/zipnerf-pytorch/assets/83005605/635db5f7-aca9-4c96-9de0-560adbe47aad">

@Ilyabasharov You can see that the circle in the ray cone is parallel to the image plane, and they actually do like this in their implementation [https://github.com/google-research/multinerf](https://github.com/google-research/multinerf)

**Comment by Ilyabasharov** [comment]
In Zip NeRF authors said that 
![github](https://github.com/SuLvXiangXin/zipnerf-pytorch/assets/25929618/7d0fcbfc-75c6-49ef-893f-443118a0bbf9)
So in your code [base_matrix](https://github.com/SuLvXiangXin/zipnerf-pytorch/blob/b1cb42943d244301a013bd53f9cb964f576b0af4/internal/render.py#LL128C5-L128C17) should be orthonormal. But first two vectors are perpendicular to `cam_dirs` and the third is `directions`. Question: `cam_dirs` must be parallel to `directions`, doesnt it?

**Comment by SuLvXiangXin** [comment]
Yeah, in the very first commit, I truely implement like that. But there comes some problems, how can I compute $\dot{r}$? While it is defined on the image plane, if I use directions as cam_dirs, that is not correct. Another solution is to use $\frac{\dot{r}}{norm(directions)}$ instead of  $\dot{r}$, but this is weird.

**Comment by Ilyabasharov** [comment]
I think all is ok in realization with `directions` because you construct orthonormal basis (length of each vector is equal 1) and than you multiply it to radius value defined and half of pixel area. $\dot{r}$ can be computed as [here](https://github.com/nerfstudio-project/nerfstudio/blob/1e205c1b5bd244dcd0c2303e2bed42d1ffc17625/nerfstudio/cameras/rays.py#L79)

**Comment by Ilyabasharov** [comment]
@SuLvXiangXin updated figure from the [paper](https://arxiv.org/pdf/2304.06706.pdf)
<img width="597" alt="Screenshot 2023-05-25 at 14 40 04" src="https://github.com/SuLvXiangXin/zipnerf-pytorch/assets/25929618/44cefb89-d9e6-4a2f-b0f9-689c12df8f3c">

**Comment by SuLvXiangXin** [comment]
@Ilyabasharov Wow, thanks for reminder, I'll checkout the updated version of the paper.
